Sources/System/FileControl.swift

```diff
@@ -0,0 +1,216 @@
+  /// `F_DUPFD_CLOEXEC`.
+  @_alwaysEmitIntoClient
+  public func duplicate(
+    minRawValue: CInt, closeOnExec: Bool
+  ) throws -> FileDescriptor {
+    let cmd: Command = closeOnExec ? .duplicateCloseOnExec : .duplicate
+    return try FileDescriptor(rawValue: fcntl(cmd, minRawValue))
+  }
+
+  // TODO: What is the firmlink thing about?
+
+  /// Get the path of the file descriptor
+  ///
+  /// - Parameters:
+  ///   - `noFirmLink`: Get the non firmlinked path of the file descriptor.
+  ///
+  /// The corresponding C functions are `fcntl` with `F_GETPATH` and
+  /// `F_GETPATH_NOFIRMLINK`.
+  public func getPath(noFirmLink: Bool = false) throws -> FilePath {
+    let cmd: Command = noFirmLink ? .getPathNoFirmLink : .getPath
+    // TODO: have a uninitialized init on FilePath / SystemString...
+    let bytes = try Array<SystemChar>(unsafeUninitializedCapacity: _maxPathLen) {
+      (bufPtr, count: inout Int) in
+      _ = try fcntl(cmd, UnsafeMutableRawPointer(bufPtr.baseAddress!))
+      // TODO: Does fcntl return the length?
+      // TODO: The below is probably the wrong formulation...
+      count = system_strlen(
+        UnsafeRawPointer(bufPtr.baseAddress!).assumingMemoryBound(to: Int8.self))
+    }
+    return FilePath(SystemString(nullTerminated: bytes))
+  }
+}
+
+// Record locking
+extension FileDescriptor {
+  /// Get record locking information.
+  ///
+  /// Get the first lock that blocks the lock description described by `lock`.
+  /// If no lock is found that would prevent this lock from being created, the
+  /// structure is left unchanged by this function call except for the lock type
+  /// which is set to F_UNLCK.
+  ///
+  /// The corresponding C function is `fcntl` with `F_GETLK`.
+  @_alwaysEmitIntoClient
+  public func getLock(blocking lock: FileLock) throws -> FileLock {
+    var copy = lock
+    try _fcntlLock(.getLock, &copy, retryOnInterrupt: false).get()
+    return copy
+  }
+
+  /// Set record locking information.
+  ///
+  /// Set or clear a file segment lock according to the lock description
+  /// `lock`.`setLock` is used to establish shared/read locks (`FileLock.read`)
+  /// or exclusive/write locks, (`FileLock.write`), as well as remove either
+  /// type of lock (`FileLock.unlock`).  If a shared or exclusive lock cannot be
+  /// set, this throws `Errno.resourceTemporarilyUnavailable`.
+  ///
+  /// If `waitIfBlocked` is set and a shared or exclusive lock is blocked by
+  /// other locks, the process waits until the request can be satisfied.  If a
+  /// signal that is to be caught is received while this calls is waiting for a
+  /// region, the it will be interrupted if the signal handler has not
+  /// specified the SA_RESTART (see sigaction(2)).
+  ///
+  /// The corresponding C function is `fcntl` with `F_SETLK`.
+  @_alwaysEmitIntoClient
+  public func setLock(
+    to lock: FileLock,
+    waitIfBlocked: Bool = false,
+    retryOnInterrupt: Bool = true
+  ) throws {
```

Comment:
Both those sounds like fine suggestions. Honestly, I think we'd go with only having the `flock()` equivalent directly on `FileDescriptor`. We can provide this functionality through the general `fcntl` fallback and not privilege POSIX-style record locks. From the man page:

> This interface follows the completely stupid semantics of System V and IEEE Std 1003.1-1988 (``POSIX.1'') that require that all locks  associated with a file for a given process are removed when any file descriptor for that file is closed by that process.  This semantic  means that applications must be aware of any files that a subroutine library may access.  For example if an application for updating the  password file locks the password file database while making the update, and then calls getpwname(3) to retrieve a record, the lock will  be lost because getpwname(3) opens, reads, and closes the password database.  The database close will release all locks that the process  has associated with the database, even if the library routine never requested a lock on the database.  Another minor semantic problem  with this interface is that locks are not inherited by a child process created using the fork(2) function.  The flock(2) interface has  much more rational last close semantics and allows locks to be inherited by child processes.  Flock(2) is recommended for applications  that want to ensure the integrity of their locks when using library routines or wish to pass locks to their children.  Note that  flock(2) and fcntl(2) locks may be safely used concurrently.

Reply:
POSIX-style record locks are crazy! Don't think we should offer them :). They are identified _per pid & VFS node_. Ie. if two threads in one process open the same file and lock distinct parts that'll succeed. So correctly, both hold a lock to distinct parts of the same file. But if one of the threads then closes its fd (which is different from the other thread's fd), then _both_ locks go away. Ie the other threads _thinks_ it still has it locked but it doesn't.